maitake/src/wait/dataqueue.rs

```diff
@@ -0,0 +1,984 @@
+            .field("waker", &self.waker)
+            .finish()
+    }
+}
+
+bitfield! {
+    #[derive(Eq, PartialEq)]
+    struct QueueState<usize> {
+        /// The queue's state.
+        const STATE: State;
+    }
+}
+
+bitfield! {
+    #[derive(Eq, PartialEq)]
+    struct WaitStateBits<usize> {
+        /// The waiter's state.
+        const STATE: WaitState;
+    }
+}
+
+/// The state of a [`Waiter`] node in a [`DataWaitQueue`].
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[repr(u8)]
+enum WaitState {
+    /// The waiter has not yet been enqueued.
+    ///
+    /// The number of times [`DataWaitQueue::wake_all`] has been called is stored
+    /// when the node is created, in order to determine whether it was woken by
+    /// a stored wakeup when enqueueing.
+    ///
+    /// When in this state, the node is **not** part of the linked list, and
+    /// can be dropped without removing it from the list.
+    Start,
+
+    /// The waiter is waiting.
+    ///
+    /// When in this state, the node **is** part of the linked list. If the
+    /// node is dropped in this state, it **must** be removed from the list
+    /// before dropping it. Failure to ensure this will result in dangling
+    /// pointers in the linked list!
+    Waiting,
+
+    /// The waiter has been woken.
+    ///
+    /// When in this state, the node is **not** part of the linked list, and
+    /// can be dropped without removing it from the list.
+    Woken,
+}
+
+/// The queue's current state.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[repr(u8)]
+enum State {
+    /// No waiters are queued, and there is no pending notification.
+    /// Waiting while the queue is in this state will enqueue the waiter;
+    /// notifying while in this state will store a pending notification in the
+    /// queue, transitioning to [`State::Woken`].
+    Empty = 0b00,
+
+    /// There are one or more waiters in the queue. Waiting while
+    /// the queue is in this state will not transition the state. Waking while
+    /// in this state will wake the first waiter in the queue; if this empties
+    /// the queue, then the queue will transition to [`State::Empty`].
+    Waiting = 0b01,
+
+    /// The queue has a stored notification. Waiting while the queue
+    /// is in this state will consume the pending notification *without*
+    /// enqueueing the waiter and transition the queue to [`State::Empty`].
+    /// Waking while in this state will leave the queue in this state.
+    Woken = 0b10,
+
+    /// The queue is closed. Waiting while in this state will return
+    /// [`Closed`] without transitioning the queue's state.
+    ///
+    /// *Note*: This *must* correspond to all state bits being set, as it's set
+    /// via a [`fetch_or`].
+    ///
+    /// [`Closed`]: crate::wait::Closed
+    /// [`fetch_or`]: core::sync::atomic::AtomicUsize::fetch_or
+    Closed = 0b11,
+}
+
+#[derive(Clone, Debug)]
+enum Wakeup {
+    Empty,
+    Waiting(Waker),
+    One,
+    Closed,
+}
```

Comment:
it seems like some of these internals could potentially be shared with `WaitQueue`?